[Messenger] Reset peak memory usage for each message

TimWolla opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -
| License       | MIT

PHP’s peak memory usage is a bit of global state that is not useful to keep in a long-running process handling individual self-contained messages, since a single high-memory message (handler) running early in a worker’s lifecycle will skew the numbers for all remaining messages processed by that worker.

By resetting the peak memory usage for each message it becomes possible to measure a given message type’s maximum memory usage more accurately, allowing to optimize hardware resources, for example by placing individual messages with handlers requiring a high memory-usage into their own transport that is executed on a larger worker instance.

As part of this change the cycle collection is also moved out of the Worker into an event-listener, since the cycle collection is not a core task of the Worker, since cycle collection would happen implicitly as well.

TimWolla committed:
[Messenger] Reset peak memory usage for each message

TimWolla commented:
I think the test failures are unrelated,k because other PRs created around the same time also fail the same tests. But please let me know if this is related to the change in `src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php`, since I'm new-ish to Symfony (contributions) :smiley:

nicolas-grekas commented:
What about resolving #59788 in the process?
Removing gc_collect_cycles might be enough.
Dunno if a middleware would be more appropriate (since @stof mentioned this way in the linked issue)

TimWolla commented:
Thanks for the link to that issue.

> Removing gc_collect_cycles might be enough.

I think having an explicit `gc_collect_cycles()` outside of the actual message processing is a good thing, since this avoids running the cycle collection while a message is currently being processed, introducing unexpected and unpredictable delays.

Though of course this amount if increased CPU usage is not a good thing either. Perhaps this is worth filing an issue with PHP itself? At least the documentation could more clearly indicate how expensive the cycle collector is.

> Dunno if a middleware would be more appropriate

I lack the Symfony experience to say whether a middleware or an event listener would be more appropriate. Please tell me what do to here, I can:

- Revert the changes for `gc_collect_cycles()`, making this PR focused on the peak memory usage reset and deferring the cycle collection solution.
- Split this into two separate event listeners, one for the reset and the other for the cycle collection.
- Split this into an event listener for the reset and a middleware for the cycle collection.
- Something else?

stof commented:
Both the reset of the peak memory usage and the cycle collection could be implemented either as middlewares or as event listeners I think. but FrameworkBundle has an existing configuration to choose which middlewares to register, while making it a listener would require introducing a new configuration setting to make it configurable. That's why I suggested a middleware initially.
Another reason is that the middleware would be usable for projects using the Messenger component without its event-dispatcher integration (the event dispatcher is optional in the Worker class). This does not matter for the framework usage as FrameworkBundle always injects the event dispatcher, but it matters for standalone usage.

TimWolla commented:
I see, thank you. I'll have a look at updating this PR to make use of middlewares and I'll report back if I have any further questions (or an updated implementation).

stof commented:
To solve https://github.com/symfony/symfony/issues/59788, I suggest that resetting the peak memory usage and collecting cycles get implemented as 2 separate middlewares.

TimWolla commented:
Yes, but running `gc_collect_cycles()` in a middleware is not particularly useful, as I've outlined in my previous comment: This kind of “clean-up” work should ideally happen after ack'ing a message and before retrieving the next one so that the message processing is not unnecessarily blocked by the cycle collection.

nicolas-grekas commented on src/Symfony/Component/Messenger/EventListener/ResetMemoryUsageListener.php:
```diff
@@ -0,0 +1,46 @@
+    {
+        return [
+            WorkerMessageReceivedEvent::class => ['resetBefore', -1024],
+            WorkerMessageFailedEvent::class => ['collectAfter', -1024],
+            WorkerMessageHandledEvent::class => ['collectAfter', -1024],
```
I suggest doing the collection only when the worker is idle. That's a good enough default to me. Ppl that need more frequent GC can still register their listener if they really want to:
```suggestion
            WorkerRunningEvent::class => ['collectAfter', -1024],
```

nicolas-grekas commented on src/Symfony/Component/Messenger/EventListener/ResetMemoryUsageListener.php:
```diff
@@ -0,0 +1,46 @@
+        memory_reset_peak_usage();
+    }
+
+    public function collectAfter(WorkerMessageHandledEvent|WorkerMessageFailedEvent $event): void
+    {
+        // Run the cycle collector after handling a message to avoid it
+        // running while the message is reserved and in processing.
+        gc_collect_cycles();
```
```suggestion
    public function collectAfter(WorkerRunningEvent $event): void
    {
        if ($event->isWorkerIdle()) {
            // Try to avoid running the GC while a message is handled by running it while the worker is idle
            gc_collect_cycles();
        }
```

nicolas-grekas commented:
Please rebase to get rid of the merge commit.
Here's my review as a diff actually:

<details>

```diff
diff --git a/src/Symfony/Component/Messenger/EventListener/ResetMemoryUsageListener.php b/src/Symfony/Component/Messenger/EventListener/ResetMemoryUsageListener.php
index c4168018bd..f5c199c530 100644
--- a/src/Symfony/Component/Messenger/EventListener/ResetMemoryUsageListener.php
+++ b/src/Symfony/Component/Messenger/EventListener/ResetMemoryUsageListener.php
@@ -12,35 +12,36 @@
 namespace Symfony\Component\Messenger\EventListener;
 
 use Symfony\Component\EventDispatcher\EventSubscriberInterface;
-use Symfony\Component\Messenger\Event\WorkerMessageFailedEvent;
-use Symfony\Component\Messenger\Event\WorkerMessageHandledEvent;
 use Symfony\Component\Messenger\Event\WorkerMessageReceivedEvent;
+use Symfony\Component\Messenger\Event\WorkerRunningEvent;
 
 /**
  * @author Tim Düsterhus <tim@tideways-gmbh.com>
  */
-class ResetMemoryUsageListener implements EventSubscriberInterface
+final class ResetMemoryUsageListener implements EventSubscriberInterface
 {
+    private bool $collect = false;
+
     public function resetBefore(WorkerMessageReceivedEvent $event): void
     {
         // Reset the peak memory usage for accurate measurement of the
         // memory usage on a per-message basis.
         memory_reset_peak_usage();
+        $this->collect = true;
     }
 
-    public function collectAfter(WorkerMessageHandledEvent|WorkerMessageFailedEvent $event): void
+    public function collectAfter(WorkerRunningEvent $event): void
     {
-        // Run the cycle collector after handling a message to avoid it
-        // running while the message is reserved and in processing.
-        gc_collect_cycles();
+        if ($event->isWorkerIdle() && $this->collect) {
+            gc_collect_cycles();
+        }
     }
 
     public static function getSubscribedEvents(): array
     {
         return [
-            WorkerMessageReceivedEvent::class => ['resetBefore', -1024],
-            WorkerMessageFailedEvent::class => ['collectAfter', -1024],
-            WorkerMessageHandledEvent::class => ['collectAfter', -1024],
+            WorkerMessageReceivedEvent::class => ['resetBefore', 1024],
+            WorkerRunningEvent::class => ['collectAfter', -1024],
         ];
     }
 }
```

</details>

chalasr commented:
Shall we backport the GC part of the patch to 6.4 to fix https://github.com/symfony/symfony/issues/59788?

nicolas-grekas commented on src/Symfony/Component/Messenger/CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@ CHANGELOG
  * Add `Symfony\Component\Messenger\Middleware\DeduplicateMiddleware` and `Symfony\Component\Messenger\Stamp\DeduplicateStamp`
  * Add `--class-filter` option to the `messenger:failed:remove` command
  * Add `$stamps` parameter to `HandleTrait::handle`
+ * Add `Symfony\Component\Messenger\EventListener\ResetMemoryUsageListener` to reset PHP’s peak memory usage for each processed message
```
```suggestion
 * Add `Symfony\Component\Messenger\EventListener\ResetMemoryUsageListener` to reset PHP's peak memory usage for each processed message
```

nicolas-grekas commented on src/Symfony/Component/Messenger/Tests/WorkerTest.php:
```diff
@@ -595,14 +596,64 @@ public function testGcCollectCyclesIsCalledOnMessageHandle()
+        $memory = str_repeat(random_int(0, 1), 4 * 1024 * 1024);
+        unset($memory);
+
+        $before = \memory_get_peak_usage();
```
```suggestion
        $before = memory_get_peak_usage();
```

nicolas-grekas commented on src/Symfony/Component/Messenger/Tests/WorkerTest.php:
```diff
@@ -595,14 +596,64 @@ public function testGcCollectCyclesIsCalledOnMessageHandle()
 
-        $gcStatus = gc_status();
+        // This should be roughly 4 MB smaller than $before.
+        $after = \memory_get_peak_usage();
```
```suggestion
        $after = memory_get_peak_usage();
```

94noni commented on src/Symfony/Component/Messenger/EventListener/ResetMemoryUsageListener.php:
```diff
@@ -0,0 +1,48 @@
+/**
+ * @author Tim Düsterhus <tim@tideways-gmbh.com>
+ */
+final class ResetMemoryUsageListener implements EventSubscriberInterface
```
wdyt to add some debug/info log here ? can it be usefull for ppl?

so they know if they want to "disable/opt out" this subscriber in their app

nicolas-grekas replied:
Let's see if anyone needs this to be configurable. I don't see why it should at the moment.

94noni replied:
got it, i was mostly commenting to [this](https://github.com/symfony/symfony/issues/59788#issuecomment-2668917134) thx

ro0NL replied:
so IIUC this middleware is not a good default to have? who will discover this if needed?

stof replied:
The difference is that `gc_collect_cycles` is potentially CPU intensive (depending on how the assigned memory looks like). On the other hand, `reset_peak_memory_usage()` is not a costly operation.

So making the usage of `gc_collect_cycles` configurable allows projects to make different choices regarding performance (trading CPU load vs more aggressive memory reduction), which is why we are doing it in 7.3 (and why we  reverted the non-configurable usage introduced before)

nicolas-grekas commented:
Thank you @TimWolla.

ro0NL commented:
@chalasr Shall we backport the GC part of the patch to 6.4 to fix https://github.com/symfony/symfony/issues/59788? was this done?

ro0NL commented:
I think im a bit a confusued by enabling gc_collect_cycles in 7.2 to then be reverted in 7.2 to then be enabled again in 7.3, which makes pinpointing things a bit difficult.
